File: backend-node/src/controllers/sosController.ts
import type { Request, Response } from "express";
import { normalizeParam } from "../utils/params.js";
import {
  createSOS,
  recordLocation,
  createPreAlert,
  cancelSOSAlert,
  getSOSStatus,
} from "../services/sosService.js";

export async function postLocation(req: Request, res: Response) {
  const touristId = normalizeParam(req.params.touristId);
  if (!touristId) {
    return res.status(400).json({ message: "Invalid tourist ID." });
  }
  const { lat, lng, accuracy, speed, heading } = req.body as {
    lat: number;
    lng: number;
    accuracy?: number;
    speed?: number;
    heading?: number;
  };
  try {
    await recordLocation(touristId, { lat, lng, accuracy, speed, heading });
    return res.status(204).send();
  } catch (error) {
    const message = (error as Error).message;
    if (message === "Tourist not found.") {
      return res.status(404).json({ message: "Tourist not found." });
    }
    throw error;
  }
}

export async function postSOS(req: Request, res: Response) {
  const touristId = normalizeParam(req.params.touristId);
  if (!touristId) {
    return res.status(400).json({ message: "Invalid tourist ID." });
  }
  const { lat, lng, message, media } = req.body as {
    lat?: number;
    lng?: number;
    message?: string;
    media?: string[];
  };
  const alert = await createSOS(touristId, lat, lng, message, media);
  return res.json({
    status: "SOS Alert initiated. Emergency response notified.",
    alertId: alert.alertId,
  });
}

export async function postPreAlert(req: Request, res: Response) {
  const touristId = normalizeParam(req.params.touristId);
  if (!touristId) {
    return res.status(400).json({ message: "Invalid tourist ID." });
  }
  const { lat, lng } = req.body as { lat?: number; lng?: number };
  const alert = await createPreAlert(touristId, lat, lng);
  return res.json({
    status: "Pre-alert registered. Monitoring initiated.",
    alertId: alert.alertId,
  });
}

export async function cancelSOS(req: Request, res: Response) {
  const alertId = Number(req.params.alertId);
  if (isNaN(alertId)) {
    return res.status(400).json({ message: "Invalid alert ID." });
  }
  const alert = await cancelSOSAlert(alertId);
  if (!alert) {
    return res.status(404).json({ message: "Alert not found." });
  }
  return res.json({ status: "Alert cancelled.", alertId: alert.alertId, alertStatus: alert.status });
}

export async function getAlertStatus(req: Request, res: Response) {
  const alertId = Number(req.params.alertId);
  if (isNaN(alertId)) {
    return res.status(400).json({ message: "Invalid alert ID." });
  }
  const status = await getSOSStatus(alertId);
  if (!status) {
    return res.status(404).json({ message: "Alert not found." });
  }
  return res.json(status);
}

File: backend-spring/src/main/java/com/safarsathi/controller/SosController.java
package com.safarsathi.controller;

import com.safarsathi.dto.LocationRequest;
import com.safarsathi.service.SosService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/action")
@RequiredArgsConstructor
public class SosController {

    private final SosService sosService;

    /**
     * POST /api/action/location/{touristId}
     */
    @PostMapping("/location/{touristId}")
    public ResponseEntity<?> postLocation(@PathVariable String touristId,
                                          @RequestBody LocationRequest request) {
        if (touristId == null || touristId.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid tourist ID."));
        }
        if (request.getLat() == null || request.getLng() == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "lat and lng required"));
        }
        try {
            sosService.recordLocation(touristId, request.getLat(), request.getLng(), request.getAccuracy());
            return ResponseEntity.noContent().build();
        } catch (RuntimeException e) {
            if ("Tourist not found.".equals(e.getMessage())) {
                return ResponseEntity.status(404).body(Map.of("message", "Tourist not found."));
            }
            throw e;
        }
    }

    /**
     * POST /api/action/sos/{touristId}
     */
    @PostMapping("/sos/{touristId}")
    public ResponseEntity<?> postSOS(@PathVariable String touristId,
                                     @RequestBody(required = false) Map<String, Double> body) {
        if (touristId == null || touristId.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid tourist ID."));
        }
        Double lat = body != null ? body.get("lat") : null;
        Double lng = body != null ? body.get("lng") : null;
        sosService.createSOS(touristId, lat, lng);
        return ResponseEntity.ok(Map.of("status", "SOS Alert initiated. Emergency response notified."));
    }
}
